### DISCUSSION TITLE
Aligning Meteor Collection methods with Mongo

### DISCUSSION THREAD

**Discussion by jamauro** [meteor/meteor#12779] [Feature Requests]
**Edit**: I'm beginning to second guess the thoughts originally expressed below. There is something nice about just having `insert`, `update`, `remove` rather than all the `*One` and `*Many` collection methods. So maybe all that is really needed is to allow `insert` to accept an array of docs and under the hood it uses `insertMany`. 

I still think that reaching for `rawCollection` should be rarely needed and seen almost akin to an anti-pattern. To that aim, adding collection methods like these would be welcomed:

```
aggregate
bulkWrite
findOneAndUpdate
``` 

```js
p.s.  I still don't like the *Async suffix so hoping that gets removed one day :) 
```

-----

**Original post for posterity**

I have a proof of concept for `insertMany` working end to end. But before I attempt to go further, I have some questions for the core team and community members who understand Meteor internals. 

First some context:
I’d like to see Meteor align with Mongo and the Node driver for Collection methods. Currently Meteor is using deprecated methods and is missing a few key methods that many people have asked for previously. Additionally, I think it’s important to keep Meteor’s philosophy of simplifying things as much as possible and reduce the need to reach for using `rawCollection()`, which is a source of confusion. 

Aligning with Mongo and the Node driver would also allow Meteor to remove the hack that suppresses deprecation warnings for insert, update, delete, which will surely cause problems at some point. 

For the primary CRUD operations — `insertOne`, `insertMany`, `updateOne`, `updateMany`, `deleteOne`, `deleteMany` — I think there are two ways to go about it:

1. Deprecate `insert`, `update`, and `remove` entirely as mongosh and the official Node driver have done. Give a grace period to migrate before removing them from the codebase. 
2. Keep `insert`, `update`, and `remove` and use them to create the other functions. For example, `insert` would be modified to also accept an array of documents and would call `insertOne` or `insertMany` as needed. Then the public facing insertOne and insertMany would basically become small helper functions that use insert. We could message to users that they shouldn’t use insert, update, and remove. We could continue allowing their use but not really advertise them and have all the guides and other documentation not use them. The benefit here would be complete backwards compatibility. The downside would be complexity and potential confusion.

Personally I think #1 is the way to go. What do you think? 

@StorytellerCZ, @zodern, @radekmie, @Grubba27, @denihs, @henriquealbert, @filipenevola I would appreciate your thoughts. 

@fredmaiaarantes, is Meteor Software interested in creating a bounty for this?

IMO, it should encompass:
```
insertOne
insertMany
updateOne
updateMany
deleteOne
deleteMany
```

I’d suggest a separate bounty or bounties for:
```
aggregate
bulkWrite
findOneAndUpdate
```

I think if there is clarity on either a) "yes, we’d like to align Meteor with Mongo and the Node driver" or b) "no, we prefer to have Meteor be different and maintain the existing methods for these reasons..." someone could theoretically get started on implementing them now and plan to have them merged after 3.0.

**Comment by filipenevola** [comment]
### Background

To provide some history: we have been aware for a long time that we are not following MongoDB Node.js driver methods, but as we knew a big migration due to Fibers was coming, we preferred to keep the complexity behind the hood and did not expose these breaking changes to Meteor users.

The risk of this approach is that Meteor Collection methods are getting more distant from MongoDB Node.js driver API, which will be very weird for newcomers.

I think we need to balance the trade-off between being friendly to newcomers and friendly to existing codebases.

Until Fibers migration is done, I would vote for friendly to existent codebases. After this period, I would be ok with more aggressive changes.

I believe after Fibers we will be able to migrate these methods almost automatically, writing codemods for each case.

Now we have too many differences to address.

### Bounty

Bounty would be a good option for this work, but different bounties for different methods would require coordination from the core team to avoid inconsistency across methods.

### Insert many

Is your code for `insertMany` already available in a PR? It would be a good starting point for discussions.

I'm also curious to see the impact of it in redis-oplog and other packages.

**Reply by jamauro** [reply]
> Until Fibers migration is done, I would vote for friendly to existent codebases. After this period, I would be ok with more aggressive changes.

IMO, a deprecation warning and extended grace period could serve as being friendly. For example you could keep the existing methods alongside the new ones until a year or so from when the new ones land. 

> I believe after Fibers we will be able to migrate these methods almost automatically, writing codemods for each case.

Nice! In that case, maybe I’ll hold off on spending more time on these. 

> would require coordination from the core team to avoid inconsistency across methods.

This is a good point. It's true for any bounty code that’s merged into core, right? Maybe the answer is contribution guidelines, some way to ensure consistent code style / formatting, etc. 

> Is your code for insertMany already available in a PR?

I could put it in a WIP PR for people to have a look at if helpful. I got it working but it needs to be cleaned up. I didn't want to spend too much time on it before starting this discussion but I was curious about what was possible. 🙂

**Comment by radekmie** [comment]
I personally would go with native names, e.g., `insertOne` or `insertMany` and make them return `Promise`s, just like the driver. It's not breaking the compatibility, since Meteor has different method names (e.g., `insert`) and as such there's no need to `*Async` them. That's what I did with `countDocuments` and `estimatedDocumentCount` in https://github.com/meteor/meteor/pull/12295.

As for `aggregate` and `findOneAndUpdate`, it's a bigger problem, since these will be _hard_ to do on the client at all. I still think that as soon as they'll become available on the server, they should also be available on the client and have the same type. But until we can truly implement them, let them just throw an error that they are server-only.

**Comment by drone1** [comment]
@jamauro are you still working on this? Fibers are long gone with Meteor 3+. I was just looking at the docs and was a bit bummed to see there is no `insertManyAsync`, etc, yet.

**Reply by jamauro** [reply]
No, there didn't seem to be much interest on this one and I recently came around to rethinking it entirely. See the edit that I made on the original post. Would be curious to hear what you and others think.

**Reply by drone1** [reply]
Do you have a link handy? Thanks

**Reply by jamauro** [reply]
I meant the `Edit` above, at the very top of this discussion :)

**Comment by jamauro** [comment]
Had some second thoughts on this one. Edited the original post accordingly.

**Reply by StorytellerCZ** [reply]
I can get behind this new version. @drone1 has a good point as well, but I think a simple `insert` vs `insertOne` and `insertMany` is easily understandable.
I would suggest to add more apis like `replace`, but if anything `aggregate` and `bulkWrite` are a must. If anything I think the biggest challenge is updating Minimongo which needs to implement all that on client. Maybe we could migrate to some library that does what Minimongo does, but is more maintained (I do believe there was one last time I looked).

**Reply by zodern** [reply]
I don't have an strong opinion on this, but I thought I would share two things:

- There's also the issue of sharing code between Meteor and non-meteor apps. I know some companies that do this, and it is complicated by Meteor and the Mongo driver having different names for the collection methods.
- The difference in the api names is understandable, but I know I've personally made mistakes before when switching between Meteor, node, the mongo shell, and blog posts or documentation. It would be nice if it is consistent.

**Reply by jamauro** [reply]
> If anything I think the biggest challenge is updating Minimongo which needs to implement all that on client

I'm not sure it would actually be that challenging but maybe I'm missing something. 

For inserting an array of docs, we could just use a for loop over an internal `._insert` since it's all in memory and the number of docs I assume will be fairly low. 

The others could use a best effort approach, e.g. `findOneAndUpdate` (if it can't find it on the client, then do nothing), or a `noop`, e.g. `aggregate`, so they can be used in isomorphic code but on the client nothing happens.

**Reply by StorytellerCZ** [reply]
@jamauro for stuff like insert it would not be an issue, but it presents new challenges when adding new APIs and extending the existing. There are still some things that don't work out of the box on the existing one.

**Comment by drone1** [comment]
i don’t think it makes sense to alter the functionality of existing
functions or semantically diverge from what the driver does. it could be
argued that doing so could create pitfalls for developers who are familiar
with mongo and are expecting alignment. just my 2 cents.
